internal/engine/wazevo/backend/isa/amd64/machine.go

```diff
@@ -132,11 +147,210 @@ func (m *machine) LowerInstr(instr *ssa.Instruction) {
+	mov := m.allocateInstr()
+	mov.asMovRR(rn.r, tmp, _64)
+
+	alu := m.allocateInstr()
+	alu.asAluRmiR(aluRmiROpcodeAdd, rm, tmp, _64)
+	m.insert(alu)
+}
+
+func (m *machine) getOperandGP(def *backend.SSAValueDefinition) operand {
+	var v regalloc.VReg
+	if def.IsFromBlockParam() {
+		v = def.BlkParamVReg
+	} else {
+		instr := def.Instr
+		if instr.Constant() {
+			panic("TODO: instr.Constant")
+		} else {
+			if n := def.N; n == 0 {
+				v = m.c.VRegOf(instr.Return())
+			} else {
+				_, rs := instr.Returns()
+				v = m.c.VRegOf(rs[n-1])
+			}
+		}
+	}
+
+	r := v
+
+	//switch inBits := def.SSAValue().Type().Bits(); {
+	//case mode == extModeNone:
+	//case inBits == 32 && (mode == extModeZeroExtend32 || mode == extModeSignExtend32):
+	//case inBits == 32 && mode == extModeZeroExtend64:
+	//	extended := m.compiler.AllocateVReg(ssa.TypeI64)
+	//	ext := m.allocateInstr()
+	//	ext.asExtend(extended, v, 32, 64, false)
+	//	m.insert(ext)
+	//	r = extended
+	//case inBits == 32 && mode == extModeSignExtend64:
+	//	extended := m.compiler.AllocateVReg(ssa.TypeI64)
+	//	ext := m.allocateInstr()
+	//	ext.asExtend(extended, v, 32, 64, true)
+	//	m.insert(ext)
+	//	r = extended
+	//case inBits == 64 && (mode == extModeZeroExtend64 || mode == extModeSignExtend64):
+	//}
+	//return operandNR(r)
+
+	return newOperandReg(r)
+}
+
+// callerGenVRegToFunctionArg is the opposite of GenFunctionArgToVReg, which is used to generate the
+// caller side of the function call.
+func (m *machine) callerGenVRegToFunctionArg(a *backend.FunctionABI, argIndex int, reg regalloc.VReg, def *backend.SSAValueDefinition, slotBegin int64) {
+	arg := &a.Args[argIndex]
+	if def != nil && def.IsFromInstr() {
+		// Constant instructions are inlined.
+		if inst := def.Instr; inst.Constant() {
+			m.InsertLoadConstant(inst, reg)
+		}
+	}
+	if arg.Kind == backend.ABIArgKindReg {
+		m.InsertMove(arg.Reg, reg, arg.Type)
+	} else {
+		// TODO: we could use pair store if there's consecutive stores for the same type.
+		//
+		// Note that at this point, stack pointer is already adjusted.
+		bits := arg.Type.Bits()
+		am := m.resolveAddressModeForOffset(arg.Offset-slotBegin, bits, rspVReg)
+		store := m.allocateInstr()
+		store.asMovRM(reg, newOperandMem(am), bits/8)
+		m.insert(store)
+	}
+}
+
+func (m *machine) callerGenFunctionReturnVReg(a *backend.FunctionABI, retIndex int, reg regalloc.VReg, slotBegin int64) {
+	r := &a.Rets[retIndex]
+	if r.Kind == backend.ABIArgKindReg {
+		m.InsertMove(reg, r.Reg, r.Type)
+	} else {
+		// TODO: we could use pair load if there's consecutive loads for the same type.
+		dst := m.resolveAddressModeForOffset(a.ArgStackSize+r.Offset-slotBegin, r.Type.Bits(), rspVReg)
+		ldr := m.allocateInstr()
+		switch r.Type {
+		case ssa.TypeI32, ssa.TypeI64:
+			ldr.asLEA(dst, reg)
+		case ssa.TypeF32, ssa.TypeF64, ssa.TypeV128:
+			panic("TODO") // ldr.asFpuLoad(operandNR(reg), amode, r.Type.Bits())
+		default:
+			panic("BUG")
+		}
+		m.insert(ldr)
```

Comment:
let's put the stack based args on hold. you can do it later.

```suggestion
		panic("TODO")
```